### PR TITLE
add details about Newton solver.

### PR DESCRIPTION
**Pull request by charlesm93**
#### Submission Checklist

- [ ] Builds locally (builds locally, but running ./build.py sometimes returns a segfault error)
- [x] Declare copyright holder and open-source license: see below

#### Summary

Add documentation for the Newton algebraic solver. Solves issue #218 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Charles Margossian, Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by charlesm93** [comment]
I only committed src files I edited. I'm assuming the doc gets built automatically.

**Comment by charlesm93** [comment]
Tagging potential reviewers: @mitzimorris, @wds15.

**Comment by charlesm93** [comment]
Do I need to review this, or can I simply merge this?

**Comment by wds15** [comment]
I will take a look

File: src/functions-reference/higher-order_functions.Rmd [hunk 1]
```diff
@@ -78,15 +81,30 @@ hybrid algorithm.
 Solves the algebraic system, given an initial guess, using the Powell
 hybrid algorithm with additional control parameters for the solver.
 
+*Note:* In future releases, the function `algebra_solver` will be deprecated
+and replaced with `algebra_solver_powell`.
```
Comment (wds15) [inline src/functions-reference/higher-order_functions.Rmd:85]: I would not make reference to a function which does not even exist at the moment. The Powell thing is not even available, right? So don't mention.

File: src/functions-reference/higher-order_functions.Rmd [hunk 2]
```diff
@@ -126,14 +144,23 @@ solver directly to the system function.
 
 #### Algorithmic details
 
-The algebraic solver uses the Powell hybrid method [@Powell:1970],
+Stan offers two algebraic solvers.
+The first one, `algebra_solver`, uses the Powell hybritd method [@Powell:1970],
```
Comment (wds15) [inline src/functions-reference/higher-order_functions.Rmd:148]: I find calling them "first" and "second" odd - it refers to historical order, yeah, but that really does not matter to the reader. Maybe instead "One solver is the Powell"... "The other solver is..."

File: src/functions-reference/higher-order_functions.Rmd [hunk 3]
```diff
@@ -126,14 +144,23 @@ solver directly to the system function.
...
 nonlinear optimization problems of the Eigen library [@Eigen:2013].
 This solver is in turn based on the algorithm developed for the
 package MINPACK-1 [@minpack:1980].
 
-The Jacobian of the solution with respect to auxiliary parameters is
+The second solver, `algebra_solver_newton`, uses Newton's method,
+also a first-order derivative based numerical solver.
+The Stan code builds on the implementation in KINSOL
+from the SUNDIALS suite [@Hindmarsh:2005].
+In general, this method works better than Powell's algorithm,
```
Comment (wds15) [inline src/functions-reference/higher-order_functions.Rmd:159]: Ui, is it that much better in general? In what terms is it better - faster, more precise or robust or even all of that?

File: src/functions-reference/higher-order_functions.Rmd [hunk 4]
```diff
@@ -126,14 +144,23 @@ solver directly to the system function.
...
+though the user should be prepared to try both solvers.
```
Comment (wds15) [inline src/functions-reference/higher-order_functions.Rmd:160]: I would refrain from encouraging to try more stuff, but ok... fine.

File: src/functions-reference/higher-order_functions.Rmd (hunk 1 above)
Comment (charlesm93) [reply src/functions-reference/higher-order_functions.Rmd:85]: We don't really justify why the Powell solver is simply called `algebra_solver`. As a reader I would find this odd. I think it's good to indicate our nomenclature is transitioning. This documentation is temporary anyways and will be updated once I put in a new PR with the renamed solver (and the variadic arguments).

File: src/functions-reference/higher-order_functions.Rmd (hunk 3 above)
Comment (charlesm93) [reply src/functions-reference/higher-order_functions.Rmd:159]: I'm not 100% sure. In the settings I tested, the Newton solver has always been faster (2 - 20 times), and returned a answer consistent with Powell's method. The examples I looked at were PK steady states and Laplace approximations. The dogleg solvers may be better when the initial guess is far from the solution.

It's possible that KINSOL is better implemented than Eigen, so the difference is an engineering one, as opposed to simply an algorithmic one.

@yizhang-yiz probably has a better understanding of the matter than I do, and I can update the documentation accordingly.

File: src/functions-reference/higher-order_functions.Rmd (hunk 4 above)
Comment (charlesm93) [reply src/functions-reference/higher-order_functions.Rmd:160]: This is poorly written. What I meant to say is that if one of the solver doesn't work well, the user should try the other solver. I've updated this:

For many problems, we find that `algebra_solver_newton` is faster  than Powell's method. If however Newton's method performs poorly, either failing to or requiring an excessively long time to converge, the user should be prepared to switch to `algebra_solver`.

**Pull request merged**